File: app/src/main/java/be/digitalia/fosdem/viewmodels/ExternalBookmarksViewModel.java
package be.digitalia.fosdem.viewmodels;

import android.app.Application;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;
import be.digitalia.fosdem.db.AppDatabase;
import be.digitalia.fosdem.model.StatusEvent;

import java.util.Arrays;

public class ExternalBookmarksViewModel extends AndroidViewModel {

	private final AppDatabase appDatabase = AppDatabase.getInstance(getApplication());
	private final MutableLiveData<long[]> bookmarkIds = new MutableLiveData<>();
	private final LiveData<PagedList<StatusEvent>> bookmarks = Transformations.switchMap(bookmarkIds,
			bookmarkIds -> new LivePagedListBuilder<>(appDatabase.getScheduleDao().getEvents(bookmarkIds), 20)
					.build());

	public ExternalBookmarksViewModel(@NonNull Application application) {
		super(application);
	}

	public void setBookmarkIds(@NonNull long[] bookmarkIds) {
		if (!Arrays.equals(bookmarkIds, this.bookmarkIds.getValue())) {
			this.bookmarkIds.setValue(bookmarkIds);
		}
	}

	public LiveData<PagedList<StatusEvent>> getBookmarks() {
		return bookmarks;
	}
}

File: app/src/main/java/be/digitalia/fosdem/viewmodels/ExternalBookmarksViewModel.kt
package be.digitalia.fosdem.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.PagingData
import androidx.paging.cachedIn
import be.digitalia.fosdem.alarms.AppAlarmManager
import be.digitalia.fosdem.db.BookmarksDao
import be.digitalia.fosdem.db.ScheduleDao
import be.digitalia.fosdem.model.StatusEvent
import be.digitalia.fosdem.utils.BackgroundWorkScope
import dagger.assisted.Assisted
import dagger.assisted.AssistedFactory
import dagger.assisted.AssistedInject
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch

class ExternalBookmarksViewModel @AssistedInject constructor(
    scheduleDao: ScheduleDao,
    private val bookmarksDao: BookmarksDao,
    private val alarmManager: AppAlarmManager,
    @Assisted private val bookmarkIds: LongArray
) : ViewModel() {

    val bookmarks: Flow<PagingData<StatusEvent>> =
        Pager(PagingConfig(20)) {
            scheduleDao.getEvents(bookmarkIds)
        }.flow.cachedIn(viewModelScope)

    fun addAll() {
        BackgroundWorkScope.launch {
            bookmarksDao.addBookmarks(bookmarkIds).let { alarmInfos ->
                alarmManager.onBookmarksAdded(alarmInfos)
            }
        }
    }

    @AssistedFactory
    interface Factory {
        fun create(bookmarkIds: LongArray): ExternalBookmarksViewModel
    }
}